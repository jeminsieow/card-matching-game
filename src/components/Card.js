import React, { useEffect, useRef } from "react";
import { View, Animated, StyleSheet, TouchableOpacity } from "react-native";
import styled from "styled-components";
import { COLOR_BLUE, COLOR_WHITE, COLOR_BLACK, CARD_WIDTH, CARD_HEIGHT } from "../utilities/constants";

const FrontText = styled.Text`
  font-size: 20px;
  color: ${COLOR_BLACK};

  ${({ flipped }) => !flipped && `color:${COLOR_WHITE};`}
`;

const BackText = styled.Text`
  font-size: 40px;
  color: ${COLOR_WHITE};
`;

const AnimatedTouchable = Animated.createAnimatedComponent(TouchableOpacity);

export default function Card({
  onPress,
  card,
  index,
  isDisabled,
  isCleared,
  isFlipped,
}) {
  const animatedValue = useRef(new Animated.Value(180)).current;

  frontInterpolate = animatedValue.interpolate({
    inputRange: [0, 180],
    outputRange: ["0deg", "180deg"],
  });

  backInterpolate = animatedValue.interpolate({
    inputRange: [0, 180],
    outputRange: ["180deg", "360deg"],
  });

  frontOpacity = animatedValue.interpolate({
    inputRange: [89, 90],
    outputRange: [1, 0],
  });

  backOpacity = animatedValue.interpolate({
    inputRange: [89, 90],
    outputRange: [0, 1],
  });

  const frontAnimatedStyle = {
    opacity: frontOpacity,
    transform: [{ rotateY: frontInterpolate }],
  };
  const backAnimatedStyle = {
    opacity: backOpacity,
    transform: [{ rotateY: backInterpolate }],
  };

  const handlePress = () => {
    if (!isCleared && !isDisabled) {
      onPress(index);
    }
  };

  useEffect(() => {
    flipCard();
  }, [isFlipped, isCleared]);

  const flipCard = () => {
    if (!isFlipped && !isCleared) {
      Animated.spring(animatedValue, {
        toValue: 180,
        friction: 8,
        tension: 10,
        useNativeDriver: true,
      }).start();
    } else {
      Animated.spring(animatedValue, {
        toValue: 0,
        friction: 8,
        tension: 10,
        useNativeDriver: true,
      }).start();
    }
  };

  return isCleared ? (
    <View>
      <Animated.View style={[styles.card, styles.frontCard]}>
        <FrontText flipped={true}>{card}</FrontText>
      </Animated.View>
    </View>
  ) : (
    <View>
      <Animated.View
        style={[frontAnimatedStyle, styles.card, styles.frontCard]}
      >
        <FrontText flipped={isFlipped}>{card}</FrontText>
      </Animated.View>
      <AnimatedTouchable
        style={[backAnimatedStyle, styles.card, styles.backCard]}
        onPress={() => handlePress()}
        disabled={isDisabled || isFlipped}
      >
        <BackText>?</BackText>
      </AnimatedTouchable>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  card: {
    width: CARD_WIDTH,
    height: CARD_HEIGHT,
    borderRadius: 10,
    borderWidth: 3,
    borderColor: COLOR_WHITE,
    alignItems: "center",
    justifyContent: "center",
  },
  frontCard: {
    backgroundColor: COLOR_WHITE,
    backfaceVisibility: "hidden",
  },
  backCard: {
    backgroundColor: COLOR_BLUE,
    position: "absolute",
  },
});
