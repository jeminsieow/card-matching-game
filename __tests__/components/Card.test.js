import React from "react";
import { create } from "react-test-renderer";
import Card from "../../src/components/Card";
import { shallow } from "enzyme";

jest.useFakeTimers();
jest.mock("react-native/Libraries/Animated/src/NativeAnimatedHelper");

describe("Card component", () => {
  it("unflipped card renders correctly", () => {
    const tree = create(
      <Card
        onPress={() => null}
        card={1}
        index={1}
        isDisabled={false}
        isCleared={false}
        isFlipped={false}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("flipped card renders correctly", () => {
    const tree = create(
      <Card
        onPress={() => null}
        card={1}
        index={1}
        isDisabled={false}
        isCleared={false}
        isFlipped={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("cleared card renders correctly", () => {
    const tree = create(
      <Card
        onPress={() => null}
        card={1}
        index={1}
        isDisabled={false}
        isCleared={true}
        isFlipped={false}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("disabled card renders correctly", () => {
    const tree = create(
      <Card
        onPress={() => null}
        card={1}
        index={1}
        isDisabled={true}
        isCleared={false}
        isFlipped={false}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("press function works correctly", () => {
    const mockFunc = jest.fn();
    const wrapper = shallow(
      <Card
        onPress={mockFunc}
        card={1}
        index={1}
        isDisabled={false}
        isCleared={false}
        isFlipped={false}
      />
    )
    const card = wrapper.find({ testID: "touchableCard" });
    card.simulate('press');
    expect(mockFunc).toHaveBeenCalled();
  });
});
