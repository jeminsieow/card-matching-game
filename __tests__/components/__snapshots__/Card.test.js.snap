// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component cleared card renders correctly 1`] = `
<View
  style={
    Object {
      "margin": 5,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#FFFFFF",
        "borderRadius": 10,
        "borderWidth": 3,
        "height": 271,
        "justifyContent": "center",
        "width": 226,
      }
    }
  >
    <Text
      flipped={true}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 20,
          },
        ]
      }
    >
      1
    </Text>
  </View>
</View>
`;

exports[`Card component disabled card renders correctly 1`] = `
<View
  style={
    Object {
      "margin": 5,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#FFFFFF",
        "borderRadius": 10,
        "borderWidth": 3,
        "height": 271,
        "justifyContent": "center",
        "opacity": -90,
        "transform": Array [
          Object {
            "rotateY": "180deg",
          },
        ],
        "width": 226,
      }
    }
  >
    <Text
      flipped={false}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 20,
          },
        ]
      }
    >
      1
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1B98F2",
        "borderColor": "#FFFFFF",
        "borderRadius": 10,
        "borderWidth": 3,
        "height": 271,
        "justifyContent": "center",
        "opacity": 91,
        "position": "absolute",
        "transform": Array [
          Object {
            "rotateY": "360deg",
          },
        ],
        "width": 226,
      }
    }
    testID="touchableCard"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 40,
          },
        ]
      }
    >
      ?
    </Text>
  </View>
</View>
`;

exports[`Card component flipped card renders correctly 1`] = `
<View
  style={
    Object {
      "margin": 5,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#FFFFFF",
        "borderRadius": 10,
        "borderWidth": 3,
        "height": 271,
        "justifyContent": "center",
        "opacity": -90,
        "transform": Array [
          Object {
            "rotateY": "180deg",
          },
        ],
        "width": 226,
      }
    }
  >
    <Text
      flipped={true}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 20,
          },
        ]
      }
    >
      1
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1B98F2",
        "borderColor": "#FFFFFF",
        "borderRadius": 10,
        "borderWidth": 3,
        "height": 271,
        "justifyContent": "center",
        "opacity": 91,
        "position": "absolute",
        "transform": Array [
          Object {
            "rotateY": "360deg",
          },
        ],
        "width": 226,
      }
    }
    testID="touchableCard"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 40,
          },
        ]
      }
    >
      ?
    </Text>
  </View>
</View>
`;

exports[`Card component unflipped card renders correctly 1`] = `
<View
  style={
    Object {
      "margin": 5,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#FFFFFF",
        "borderRadius": 10,
        "borderWidth": 3,
        "height": 271,
        "justifyContent": "center",
        "opacity": -90,
        "transform": Array [
          Object {
            "rotateY": "180deg",
          },
        ],
        "width": 226,
      }
    }
  >
    <Text
      flipped={false}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 20,
          },
        ]
      }
    >
      1
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1B98F2",
        "borderColor": "#FFFFFF",
        "borderRadius": 10,
        "borderWidth": 3,
        "height": 271,
        "justifyContent": "center",
        "opacity": 91,
        "position": "absolute",
        "transform": Array [
          Object {
            "rotateY": "360deg",
          },
        ],
        "width": 226,
      }
    }
    testID="touchableCard"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 40,
          },
        ]
      }
    >
      ?
    </Text>
  </View>
</View>
`;
